### PR TITLE
add ReceiveFromAsync and SendToAsync with SocketAddress overload

### LINKED ISSUE #78448: Debug.Assert when using ReceiveFrom on UnixDomainSocket
I was adding new test like 
```c#
            using (Socket server = new Socket(AddressFamily.Unix, SocketType.Dgram, ProtocolType.Unspecified))
            {
                server.Bind(new UnixDomainSocketEndPoint(serverAddress));
                using (Socket client = new Socket(AddressFamily.Unix, SocketType.Dgram, ProtocolType.Unspecified))
                {
                    byte[] data = Encoding.ASCII.GetBytes(nameof(UnixDomainSocketEndPoint_SendToReceiveFrom_CorrectEndPoint));
                    // Bind the client.
                    client.Bind(new UnixDomainSocketEndPoint(clientAddress));

                    var sender = new UnixDomainSocketEndPoint(GetRandomNonExistingFilePath());
                    EndPoint senderRemote = (EndPoint)sender;
                    client.SendTo(data, server.LocalEndPoint);

                    byte[] buffer = new byte[data.Length * 2];
                    int receivedBytes = server.ReceiveFrom(buffer, ref senderRemote);
                }
            }
```

and it will produce core with 
```
   Starting:    System.Net.Sockets.Tests (parallel test collections = off, max threads = 4)
      System.Net.Sockets.Tests.UnixDomainSocketTest.UnixDomainSocketEndPoint_SendToReceiveFrom_CorrectEndPoint(abstractAddress: False) [STARTING]
  dotnet: /home/furt/github/wfurt-runtime/src/native/libs/System.Native/pal_networking.c:1407: int32_t SystemNative_ReceiveMessage(intptr_t, MessageHeader *, int32_t, int64_t *): Assertion `(int32_t)header.msg_namelen <= messageHeader->SocketAddressLen' failed.
  /home/furt/github/wfurt-runtime/artifacts/bin/System.Net.Sockets.Tests/Debug/net7.0-unix/RunTests.sh: line 168: 79756 Aborted                 (core dumped) "$RUNTIME_PATH/dotnet" exec --runtimeconfig System.Net.Sockets.Tests.runtimeconfig.json --
```

While `sockaddr_un` has fixed size the actual `path` can vary
```c
          struct sockaddr_un {
               sa_family_t sun_family;               /* AF_UNIX */
               char        sun_path[108];            /* Pathname */
           };
```
We seems to serialize only the actual size of the `EndPoint` so unless caller always creates string with  maximum platform size they may have problems. (production code truncates the address) 

Since the `EndPoint` is passed in as `ref` to be filled in it should not matter what the endpoint was. It would be nice if one could simply create empty `UnixDomainSocketEndPoint` from `string.Empty` but that fails in constructor.

### LINKED ISSUE #87397: [API Proposal]: add overloads with SocketAddress to Socket's SendTo and ReceiveFrom
### Background and motivation

This is subset from #30797. There is long discussion about how to make sending and receiving UDP more efficient. Current overloads allocate a lot for each packet and that kills performance and it puts lot of pressure on GC. Garbage collection has  significant impact on scenarios that are sensitive to timing - like multiplayer Unity games.  

### API Proposal

```csharp
namespace System.Net.Sockets;

public class Socket
{
+  public int ReceiveFrom(Span<byte> buffer, SocketFlags socketFlags, SocketAddress socketAddress);
+  public ValueTask<int> ReceiveFromAsync(Memory<byte> buffer, SocketFlags socketFlags, SocketAddress socketAddress, CancellationToken cancellationToken = default);

+. public int SendTo(ReadOnlySpan<byte> buffer, SocketFlags socketFlags, SocketAddress socketAddress);
+. public ValueTask<int> SendToAsync(ReadOnlyMemory<byte> buffer, SocketFlags socketFlags, SocketAddress socketAddress, CancellationToken cancellationToken = default);
}
```

Instead of allocating and returning EndPoint, we can reuse `SocketAddress` and we would update it in place. Ideally using API proposed in https://github.com/dotnet/runtime/issues/86872 but in worst case the `SocketAddress` is already writable and we can play some internal tricks. 

### API Usage

example of simple server that would process request - inspired by DNS server linked in #30797

```csharp
  SocketAddress sa = new SocketAddress(AddressFamily.InterNetworkV6);
  while (doWork)
  {
      int receivedLength = s.ReceiveFrom(buffer, SocketFlags.None, sa);
      response = processRequest(buffer.Slice(0, receivedLength);
      s.SendTo(response, SocketFlags.None, sa);
  }
```

The async version would be basically same with addition of `CancellationToken`. Unlike `ReceiveFromAsync ` that used `EndPoint` and returns `SocketReceiveFromResult` we would return `ValueTask<int>` just like the synchronous version and we would update `socketAddress` in-place.

### Alternative Designs

_No response_

### Risks

This is low-level API intended for experienced users. While there is nothing particularly dangerous one should be very careful about life cycle of the `socketAddress` as it can be overwritten on subsequent use.

### PR DESCRIPTION
**Pull request by wfurt**
contributes to https://github.com/dotnet/runtime/issues/30797
fixes #87397

This is the last big step. There is no longer  `Internals.SocketAddress`.
That allows to use provided `SocketAddress` without any additional processing. 

I realized that the `receivedSocketAddress` from #88970 did not sue the approved name -> this fixes it makes sync & async same.

There is probably more room for cleanup but this still should be much better - even for the legacy overloads.  

also fixes #78448

### CODE REVIEW

**Comment by stephentoub** [comment]
> contributes to https://github.com/dotnet/runtime/issues/30797

What's more to do on that issue?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 1]
```diff
@@ -203,8 +200,9 @@ private unsafe Socket(SafeSocketHandle handle, bool loadPropertiesFromHandle)
                                             break;
 
                                         case AddressFamily.Unix:
-                                            socketAddress = new Internals.SocketAddress(AddressFamily.Unix, buffer.Slice(0, bufferLength));
-                                            _remoteEndPoint = new UnixDomainSocketEndPoint(IPEndPointExtensions.GetNetSocketAddress(socketAddress));
+                                            //socketAddress = new SocketAddress(AddressFamily.Unix, buffer.Slice(0, bufferLength));
+                                            //_remoteEndPoint = new UnixDomainSocketEndPoint(IPEndPointExtensions.GetNetSocketAddress(socketAddress));
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:204]: What's the purpose of the commented out code?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 2]
```diff
@@ -1892,27 +1879,28 @@ public int ReceiveFrom(Span<byte> buffer, SocketFlags socketFlags, ref EndPoint
...
+        /// <param name="receivedAddress">An <see cref="SocketAddress"/>, that will be updated with value of the remote peer.</param>
         /// <returns>The number of bytes received.</returns>
         /// <exception cref="ArgumentNullException"><c>remoteEP</c> is <see langword="null" />.</exception>
         /// <exception cref="SocketException">An error occurred when attempting to access the socket.</exception>
         /// <exception cref="ObjectDisposedException">The <see cref="Socket"/> has been closed.</exception>
-        public int ReceiveFrom(Span<byte> buffer, SocketFlags socketFlags, SocketAddress receivedSocketAddress)
+        public int ReceiveFrom(Span<byte> buffer, SocketFlags socketFlags, SocketAddress receivedAddress)
         {
             ThrowIfDisposed();
+            ArgumentNullException.ThrowIfNull(receivedAddress, nameof(receivedAddress));
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1890]: Please ensure we have tests for all argument validation across all the new methods.

**Review by stephentoub — APPROVED** [review]
Can you share perf numbers?

**Comment by wfurt** [comment]
> > contributes to #30797
> 
> What's more to do on that issue?

nothing for 8.0. Plan to write some summary there and either keep it if we want to expose it on EventArg for consistency,  open separate issue for it or close it with comment and wait if there is need for it.

File: src/libraries/Common/src/Interop/Unix/System.Native/Interop.Bind.cs [hunk 3]
```diff
@@ -10,10 +10,10 @@ internal static partial class Interop
     internal static partial class Sys
     {
         [LibraryImport(Libraries.SystemNative, EntryPoint = "SystemNative_Bind")]
-        private static partial Error Bind(SafeHandle socket, ProtocolType socketProtocolType, ReadOnlySpan<byte> socketAddress, int socketAddressLen);
+        private static partial Error Bind(SafeHandle socket, int socketProtocolType, ReadOnlySpan<byte> socketAddress, int socketAddressLen);
```
[the remark this answers is not in the document]
Comment (wfurt) [reply src/libraries/Common/src/Interop/Unix/System.Native/Interop.Bind.cs]: It failed to compile at some point. But it is working now without change so I roll it back.

File: src/libraries/Common/src/System/Net/SocketAddressExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+        public static bool Equals(this SocketAddress socketAddress, EndPoint? endPoint)
+        {
+            if (endPoint is IPEndPoint ipe)
+            {
+                if (socketAddress.Family == endPoint.AddressFamily)
+                {
+                    return ipe.Equals(socketAddress.Buffer.Span);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (wfurt) [reply src/libraries/Common/src/System/Net/SocketAddressExtensions.cs]: updated.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs [hunk 5]
```diff
@@ -636,11 +672,41 @@ public ValueTask<int> SendToAsync(ReadOnlyMemory<byte> buffer, SocketFlags socke
...
+        /// <summary>
+        /// Sends data to the specified remote host.
+        /// </summary>
+        /// <param name="buffer">The buffer for the data to send.</param>
+        /// <param name="socketFlags">A bitwise combination of SocketFlags values that will be used when sending the data.</param>
+        /// <param name="socketAddress">The remote host to which to send the data.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used to cancel the asynchronous operation.</param>
+        /// <returns>An asynchronous task that completes with the number of bytes sent.</returns>
+        public ValueTask<int> SendToAsync(ReadOnlyMemory<byte> buffer, SocketFlags socketFlags, SocketAddress socketAddress, CancellationToken cancellationToken = default)
+        {
```
[the remark this answers is not in the document]
Comment (wfurt) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs:695]: added.

**Comment by wfurt** [comment]
This should be ready for another review  round @stephentoub 
I also added tests and fixes for #78448 since that is the same area.
I'll get perf comparison today

File: src/libraries/Common/src/System/Net/SocketProtocolSupportPal.Windows.cs [hunk 6]
```diff
@@ -10,13 +10,14 @@ internal static partial class SocketProtocolSupportPal
...
         private static bool IsSupported(AddressFamily af)
         {
+            const int StreamSocketType = 1;
             Interop.Winsock.EnsureInitialized();
 
             IntPtr INVALID_SOCKET = (IntPtr)(-1);
             IntPtr socket = INVALID_SOCKET;
             try
             {
-                socket = Interop.Winsock.WSASocketW(af, DgramSocketType, 0, IntPtr.Zero, 0, (int)Interop.Winsock.SocketConstructorFlags.WSA_FLAG_NO_HANDLE_INHERIT);
```
Comment (wfurt) [inline src/libraries/Common/src/System/Net/SocketProtocolSupportPal.Windows.cs:19]: This broke `Socket.OSSupportsUnixDomainSockets` as Windows only support Stream for UDS.

**Comment by wfurt** [comment]
|                                         Method |        Job |         Toolchain |           Mean |        Error |        StdDev |         Median |            Min |            Max | Ratio | RatioSD |     Gen0 | Allocated | Alloc Ratio |
|----------------------------------------------- |----------- |------------------ |---------------:|-------------:|--------------:|---------------:|---------------:|---------------:|------:|--------:|---------:|----------:|------------:|
|                             ConnectAcceptAsync | Job-ZDTGEN |   \PR\corerun.exe |       761.8 us |    315.49 us |     363.32 us |       513.1 us |       471.6 us |     1,547.8 us |  1.49 |    0.75 |        - |    1344 B |        1.00 |
|                             ConnectAcceptAsync | Job-VRCVXQ | \main\corerun.exe |       502.9 us |     15.49 us |      15.90 us |       509.7 us |       481.3 us |       522.3 us |  1.00 |    0.00 |        - |    1345 B |        1.00 |
|                                                |            |                   |                |              |               |                |                |                |       |         |          |           |             |
|                 SendAsyncThenReceiveAsync_Task | Job-ZDTGEN |   \PR\corerun.exe |   104,494.1 us |  2,054.93 us |   2,018.22 us |   104,477.0 us |   100,835.1 us |   108,956.7 us |  1.00 |    0.04 |        - |         - |          NA |
|                 SendAsyncThenReceiveAsync_Task | Job-VRCVXQ | \main\corerun.exe |   104,136.5 us |  3,092.13 us |   3,436.90 us |   102,974.0 us |    99,681.8 us |   112,911.4 us |  1.00 |    0.00 |        - |         - |          NA |
|                                                |            |                   |                |              |               |                |                |                |       |         |          |           |             |
|                 ReceiveAsyncThenSendAsync_Task | Job-ZDTGEN |   \PR\corerun.exe |   685,473.6 us | 13,392.59 us |  14,329.93 us |   682,778.8 us |   660,879.8 us |   712,324.7 us |  0.96 |    0.15 |        - |         - |          NA |
|                 ReceiveAsyncThenSendAsync_Task | Job-VRCVXQ | \main\corerun.exe |   728,471.7 us | 78,202.80 us |  90,058.46 us |   713,186.3 us |   510,999.9 us |   885,721.3 us |  1.00 |    0.00 |        - |         - |          NA |
|                                                |            |                   |                |              |               |                |                |                |       |         |          |           |             |
| SendAsyncThenReceiveAsync_SocketAsyncEventArgs | Job-ZDTGEN |   \PR\corerun.exe |   102,279.1 us |  1,540.97 us |   1,366.03 us |   102,311.2 us |   100,069.7 us |   104,280.6 us |  0.96 |    0.05 |        - |         - |          NA |
| SendAsyncThenReceiveAsync_SocketAsyncEventArgs | Job-VRCVXQ | \main\corerun.exe |   104,810.3 us |  4,388.80 us |   5,054.15 us |   102,546.4 us |    98,808.3 us |   113,480.5 us |  1.00 |    0.00 |        - |         - |          NA |
|                                                |            |                   |                |              |               |                |                |                |       |         |          |           |             |
| ReceiveAsyncThenSendAsync_SocketAsyncEventArgs | Job-ZDTGEN |   \PR\corerun.exe |   674,034.8 us | 44,428.74 us |  47,538.26 us |   684,631.1 us |   486,587.7 us |   698,983.0 us |  0.94 |    0.08 |        - |         - |          NA |
| ReceiveAsyncThenSendAsync_SocketAsyncEventArgs | Job-VRCVXQ | \main\corerun.exe |   719,407.7 us | 29,227.37 us |  33,658.29 us |   702,072.3 us |   683,461.1 us |   782,602.0 us |  1.00 |    0.00 |        - |         - |          NA |
|                                                |            |                   |                |              |               |                |                |                |       |         |          |           |             |
|           ReceiveFromAsyncThenSendToAsync_Task | Job-ZDTGEN |   \PR\corerun.exe | 1,587,406.7 us | 73,418.09 us |  84,548.38 us | 1,585,315.7 us | 1,407,020.1 us | 1,751,406.0 us |  0.99 |    0.08 |        - | 4321064 B |        0.75 |
|           ReceiveFromAsyncThenSendToAsync_Task | Job-VRCVXQ | \main\corerun.exe | 1,610,837.5 us | 96,790.05 us | 103,564.27 us | 1,616,955.2 us | 1,373,973.8 us | 1,811,356.2 us |  1.00 |    0.00 |        - | 5761064 B |        1.00 |
|                                                |            |                   |                |              |               |                |                |                |       |         |          |           |             |
|                          SendToThenReceiveFrom | Job-ZDTGEN |   \PR\corerun.exe |   164,729.9 us |  1,880.36 us |   1,758.89 us |   164,669.6 us |   162,245.6 us |   168,004.6 us |  0.98 |    0.02 | 500.0000 | 4320440 B |        0.75 |
|                          SendToThenReceiveFrom | Job-VRCVXQ | \main\corerun.exe |   168,548.9 us |  3,004.79 us |   2,810.68 us |   167,155.2 us |   164,936.9 us |   173,277.2 us |  1.00 |    0.00 |        - | 5761408 B |        1.00 |


shows ~ 25% allocation reduction in legacy SendTo/ReceiveFrom. We can still probably do better - tabling for 9.0

**Pull request merged**